File: java/JavaStreams/src/main/java/pl/sebcel/livecoding/javastreams/salesservice/SalesService.java
package pl.sebcel.livecoding.javastreams.salesservice;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class SalesService {

	/**
	 * Returns top customers per country
	 * 
	 * If country has less than topN customers, return them all
	 * For every customer should return sum of amounts
	 * Must throw NullPointerException if list of orders is null
	 * Must throw IllegalArgumentException if topN is negative
	 * Customers per country must be returned in descending order
	 * 
	 * @param orders list of orders
	 * @param topN number of customers to return per country
	 * @return list of top customers with sums of their orders per country
	 */
	public Map<String, List<CustomerExpenses>> getTopCustomersPerCountry(List<Order> orders, int topN) {
		Objects.requireNonNull(orders);
		if (topN < 0) {
			throw new IllegalArgumentException();
		}
		
		// group by country   -> Map<String, List<Order>>
		// inside country
		//   customer-amount aggregate by amount (sum)   ->     List<Order> -> Map<String, BigDecimal> (aggregation, sum)
		//   sort by aggregated amount desc
		//   take topN
		
		
		return orders
				.stream()
				.collect(Collectors.groupingBy(
						Order::country,						// Map<String, List<Order>>
						Collectors.collectingAndThen(
							Collectors.toMap(
									Order::customerId,			// Map<String, Map<String, List<Order>>>
									Order::amount,
									BigDecimal::add),
							x -> x.entrySet()
									.stream()
									.sorted(Map.Entry.<String, BigDecimal>comparingByValue().reversed())
									.limit(topN)
									.map(e -> new CustomerExpenses(e.getKey(), e.getValue())).toList() // Map<String, List<CustomerExpenses>>
						)));  
				
	}
	
}

File: csharp/LINQtoObjects/SalesService/SalesService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LiveCodingExercises.LINQtoObjects.SalesService
{
    public class SalesService
    {
        /// <summary>
        /// Return top customers per country with sums of their amounts
        /// <para>If country has less than topN customers, return them all</para>
        /// <para>For every customer should return sum of amounts</para>
        /// <para>Must throw ArgumentNullException if list of orders is null</para>
        /// <para>Must throw ArgumentOutOfRangeException if topN is negative</para>
        /// <para>Customers per country must be returned in descending order</para>
        /// </summary>
        /// <param name="orders">list of orders</param>
        /// <param name="topN">maximum numer of customers to return per country</param>
        /// <returns>list of top customers per country with sums of their amounts</returns>
        /// <exception cref="NotImplementedException"></exception>
        public IDictionary<string, IEnumerable<CustomerExpenses>> GetTopCustomersPerCountry(IEnumerable<Order> orders, int topN)
        {
            ArgumentOutOfRangeException.ThrowIfNegative(topN);
            ArgumentNullException.ThrowIfNull(orders);

            return orders
                .GroupBy(o => new { o.Country, o.CustomerId })   // IGrouping<(string, string), Order>
                .Select(g => new { g.Key.Country, g.Key.CustomerId, Sum = g.Sum(o => o.Amount) })  // IEnumerable<(string, string, decimal)>
                .GroupBy(h => h.Country) // IGrouping<string, <(string, string, decimal)>>
                .ToDictionary(
                    i => i.Key,
                    i => i.OrderByDescending(i => i.Sum)
                            .Take(topN)
                            .Select(k => new CustomerExpenses(k.CustomerId, k.Sum)));
        }
    }
}
